interp/notation_term.mli

```diff
@@ -86,11 +93,21 @@ type notation_binder_source =
   (* Parsed as open or closed binder: This accepts ident, _, quoted pattern, etc. *)
   | NtnBinderParsedAsBinder
 
-type notation_var_instance_type =
-  | NtnTypeConstr
-  | NtnTypeConstrList
-  | NtnTypeBinder of notation_binder_source
-  | NtnTypeBinderList of notation_binder_source
+type notation_var_constr_kind =
+  | NtnAlwaysConstr (* In anticipation of supporting PatCast in pattern *)
+  | NtnConstrForConstrAndPatternForPattern
+
+(** Type of variables in notation interpretations, remembering how it is parsed;
+    Note: when a notation variable appears both as term and as binder, its type is Pattern *)
+type notation_var_kind =
+  | NtnTypeVarConstr of notation_var_constr_kind
+  | NtnTypeVarPattern of notation_binder_source
+  | NtnTypeVarBinders of notation_binder_source
+
+type notation_var_type =
+  | NtnTypeVar of (extended_subscopes * notation_var_binders) * notation_var_kind
+  | NtnTypeVarList of notation_var_type (* list A *)
+  | NtnTypeVarTuple of notation_var_type list (* A1 * ... * An *)
```

Comment:
We should probably document all of this a bit more. For instance why does NtnTypeVarConstr need an additional NtnAlwaysConstr?